### PR TITLE
[wip] update protos to proto3 and clean up dialyzer warnings

### PR DESCRIPTION
**Pull request by tsloughter**
This updates to proto3 and fixes a few other dialyzer warnings.

The main dialyzer related changes were just one in `hex_http`, `hex_erl_tar.hrl` and `hex_registry`

I removed the gpb option to target 17. I don't think it is necessary, but we'll see when the tests run on 17 in travis :), except for type specs, which may as well only need to be correct for the latest otp and dialyzed only on otp-21.

I had to remove `{verify, always}` and replace it with optional. I opened an issue with gpb because the issue really confuses me https://github.com/tomas-abrahamsson/gpb/issues/149

It is marked `wip` because the `hex_tarball_SUITE` `disk_test` checksum comparison currently fails. It doesn't seem protos are included in any way in the tarballs, right? I can't find what would have changed to cause this, so I am leaving it failing until I do and can fix it.

### CODE REVIEW

**Comment by tsloughter** [comment]
Added back `target_erlang_version` set to `20` so the stacktrace stuff works correctly on older versions. Not sure why it doesn't just have the default be the one using `get_stacktrace` and ones with `OTP_RELEASE` defined use the pattern...

**Comment by tsloughter** [comment]
Well, nevermind. Without some tweaks to gpb we can not get this to work with versions as far back as 17 and pass dialyzer on 21.

Since that turns out to be the case I've set `verify` back to `always` like it was originally and `target` back to 17 (it couldn't build at all when gpb was adding `-dialyzer` attributes before functions, which it does for versions after 17).

**Comment by ericmj** [comment]
> It is marked wip because the hex_tarball_SUITE disk_test checksum comparison currently fails. It doesn't seem protos are included in any way in the tarballs, right? I can't find what would have changed to cause this, so I am leaving it failing until I do and can fix it.

The PR is marked as [wip] but it seems like all tests are passing now. Is there anything left to do?

File: src/hex_http.erl [hunk 1]
```diff
@@ -8,14 +8,14 @@
 -type method() :: get | post | put | patch | delete.
 -type status() :: non_neg_integer().
 -type headers() :: #{binary() => binary()}.
--type body() :: nil.
+-type body() :: binary() | {string(), binary()} | nil.
```
Comment (ericmj) [inline src/hex_http.erl]: What is `{string(), binary()}` used for? /cc @wojtekmach

Comment (tsloughter) [reply src/hex_http.erl]: Body can be the tuple `{ContentType, Body}`. It is used in `hex_http_httpc`.

**Comment by tsloughter** [comment]
@ericmj so I don't know why it passes is the issue. The tarball checksum test fails for me locally.

Can you or @wojtekmach check if it fails for you locally? Curious to know if its something with my system or it is broken in such a way that it only passes in travis.

**Comment by mrallen1** [comment]
@tsloughter For whatever it's worth, `rebar3 ct` fails for me too on 21.0.1 with a checksum error. So it's not just you.

**Comment by tsloughter** [comment]
@mrallen1 cool, thanks. I'll keep poking at it.

**Comment by wojtekmach** [comment]
@tsloughter sorry for dropping the ball on this; could you rebase against master?

We've just committed this on master: https://github.com/hexpm/hex_core/commit/28acd1879f4e31d3402d40b129a37dae2f7068e9 which seems related to what you ran into.

**Comment by wojtekmach** [comment]
@tsloughter actually never mind, I'll take care of it and merge manually!

**Comment by wojtekmach** [comment]
Closing in favour of https://github.com/hexpm/hex_core/pull/46

**Pull request closed without merging**